README.MD
# NET-CAT

## Authors:

atemerzh and asundeto

## Description:

This project consists on recreating the NetCat in a Server-Client Architecture that can run in a server mode on a specified port listening for incoming connections, and it can be used in client mode, trying to connect to a specified port and transmitting information to the server.

* NetCat, nc system command, is a command-line utility that reads and writes data across network connections using TCP or UDP. It is used for anything involving TCP, UDP, or UNIX-domain sockets, it is able to open TCP connections, send UDP packages, listen on arbitrary TCP and UDP ports and many more.

### The project must have the following features :

* TCP connection between server and multiple clients (relation of 1 to many).
* A name requirement to the client.
* Control connections quantity.
* Clients must be able to send messages to the chat.
* Do not broadcast EMPTY messages from a client.
* Messages sent, must be identified by the time that was sent and the user name of who sent the message, example : [2020-01-20 15:48:41][client.name]:[client.message]
* If a Client joins the chat, all the previous messages sent to the chat must be uploaded to the new Client.
* If a Client connects to the server, the rest of the Clients must be informed by the server that the Client joined the group.
* If a Client exits the chat, the rest of the Clients must be informed by the server that the Client left.
* All Clients must receive the messages sent by other Clients.
* If a Client leaves the chat, the rest of the Clients must not disconnect.
* If there is no port specified, then set as default the port 8989. Otherwise, program must respond with usage message: [USAGE]: ./TCPChat $port

## Usage: how to run

student@ALEM~/net-cat$ **go run .**

Listening on the port :8989

**or** 

student@ALEM~/net-cat$ **go run . $port**

Listening on the port :$port

* Open the terminal

* Type **nc localhost $port**

* If you have done everything correctly, the TCP-Chat logo will appear!

* Example:
Welcome to TCP-Chat!
         _nnnn_
        dGGGGMMb
       @p~qp~~qMb
       M|@||@) M|
       @,----.JM|
      JS^\__/  qKL
     dZP        qKRb
    dZP          qKKb
   fZP            SMMb
   HZM            MMMM
   FqM            MMMM
 __| ".        |\dS"qML
 |    `.       | `' \Zq
_)      \.___.,|     .'
\____   )MMMMMP|   .'
     `-'       `--'
[ENTER YOUR NAME]:

* Repeat with another terminals

history.txt
lfdf has joined.
[07-01-2023 20:57:51][lfdf]: sdfbgbhn
hm has joined.
[07-01-2023 20:58:12][hm]: bv
lfdf has left.
gh has joined.


main.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"log"
	"net"
	"os"
	"time"
)

var (
	clientsStr []string
	clientsNet []net.Conn
	leaving    = make(chan message)
	messages   = make(chan message)
	logo       = ReadWholeFile("./logo.txt")
	file       os.File
	boly       bool
	start      bool
)

type message struct {
	text    string
	address string
}

func main() {
	port := "8989"
	if len(os.Args) == 2 {
		port = os.Args[1]
	} else if len(os.Args) > 2 {
		fmt.Println("[USAGE]: ./TCPChat $port")
		return
	}
	// mux := http.NewServeMux()
	// err := http.ListenAndServe(":"+port, mux)
	// log.Fatal(err)
	listen, err := net.Listen("tcp", "localhost:"+port)
	if err != nil {
		log.Println("Incorrect port, try another one!")
		return
	}
	file1, err := os.Create("history.txt")
	if err != nil {
		fmt.Println(err)
		return
	}
	file = *file1
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Listening on the port:", port)
	go broadcaster()
	for {
		conn, err := listen.Accept()
		if err != nil {
			log.Print(err)
			continue
		}
		go handle(conn)
	}
}

func handle(cl net.Conn) {
	var login bool
	if checkCnt(cl) {
		cl.Close()
		return
	}
	cl.Write([]byte(logo))
	cl.Write([]byte("Enter your name: "))
	input := bufio.NewScanner(cl)
	x := 0
	for input.Scan() {
		if x == 0 {
			name := string(input.Text())
			if checkUserName(name, cl) {
				continue
			}
			clientsStr = append(clientsStr, name)
			clientsNet = append(clientsNet, cl)
			history, err := ioutil.ReadFile(file.Name())
			if err != nil {
				fmt.Println("Failed to open history:", err)
				return
			}
			cl.Write(history)
			messages <- newMessage(" has joined.", cl, false)
			x++
			login = true
			continue
		}
		inpTxt := input.Text()
		if checkMsg(inpTxt, cl) {
			continue
		}
		messages <- newMessage(inpTxt, cl, true)
		x++
	}
	if login {
		leaving <- newMessage(" has left.", cl, false)

		deleteClient(cl)

		cl.Close()
	} else {
		cl.Close()
	}
}

func deleteClient(cl net.Conn) {
	var newClientsStr []string
	var newClientsNet []net.Conn
	for i := 0; i < len(clientsNet); i++ {
		if clientsNet[i] != cl {
			newClientsNet = append(newClientsNet, clientsNet[i])
			newClientsStr = append(newClientsStr, clientsStr[i])
		}
	}
	clientsStr, clientsNet = newClientsStr, newClientsNet
}

func saveHistory(s string, h *os.File) {
	if _, err := h.WriteString(s); err != nil {
		log.Fatal()
	}
}

func checkMsg(s string, cl net.Conn) bool {
	if s == "" {
		name := "[" + clientsStr[getId(cl)] + "]"
		cl.Write([]byte(getTime(name)))
		return true
	}
	return false
}

func checkCnt(cl net.Conn) bool {
	if len(clientsStr) >= 2 {
		cl.Write([]byte("Server is full! Try again latter."))
		return true
	}
	return false
}

func getId(cl net.Conn) int {
	for i := 0; i < len(clientsNet); i++ {
		if cl == clientsNet[i] {
			return i
		}
	}
	return 0
}

func checkUserName(s string, cl net.Conn) bool {
	if s == "" {
		cl.Write([]byte("Please enter correct name!\nEnter your name: "))
		return true
	}
	for i := 0; i < len(clientsStr); i++ {
		if s == clientsStr[i] {
			cl.Write([]byte("Username exists! Try another username.\nEnter your name: "))
			return true
		}
	}
	return false
}

func newMessage(msg string, conn net.Conn, b bool) message {
	addr := conn.RemoteAddr().String()
	name := clientsStr[getId(conn)]
	if b {
		name = "[" + name + "]"
		name = getTime(name)
		boly = false
	} else {
		boly = true
	}
	return message{
		text:    name + msg,
		address: addr,
	}
}

func getTime(name string) string {
	return fmt.Sprintf("\r[%s]%s: ", time.Now().Format("01-02-2006 15:04:05"), name)
}

func broadcaster() {
	for {
		select {
		case msg := <-messages:
			for _, conn := range clientsNet {
				name := "[" + clientsStr[getId(conn)] + "]"
				name = getTime(name)
				if msg.address == conn.RemoteAddr().String() {
					conn.Write([]byte(name))
					continue
				}
				conn.Write([]byte("\n" + msg.text + "\n" + name))
			}
			if boly {
				if !start {
					saveHistory(msg.text+"\n", &file)
					start = true
					continue
				}
				saveHistory(msg.text+"\n", &file)
			} else {
				saveHistory(msg.text+"\n", &file)
			}
		case msg := <-leaving:
			for _, conn := range clientsNet {
				conn.Write([]byte("\n" + msg.text + "\n"))
				evrName := "[" + clientsStr[getId(conn)] + "]"
				evrName = getTime(evrName)
				conn.Write([]byte(evrName))
			}
			saveHistory(msg.text+"\n", &file)
		}
	}
}

func ReadWholeFile(logo string) string {
	contents, err := ioutil.ReadFile(logo)
	if err != nil {
		fmt.Println(err.Error())
	}
	if len(string(contents)) == 0 {
		fmt.Println("Error! Empty parametre!")
		return "ERROR"
	}
	return string(contents)
}
